src/java.base/share/classes/java/util/zip/ZipFile.java

```diff
@@ -1400,16 +1403,34 @@ private static class Key {
             File file;
             final boolean utf8;
 
-            public Key(File file, BasicFileAttributes attrs, ZipCoder zc) {
+            public Key(File file, BasicFileAttributes attrs, ZipCoder zc)
+                    throws IOException {
                 this.attrs = attrs;
-                this.file = file;
+                Object fk = attrs.fileKey();
+                this.file = (fk != null) ? file : getCanonicalFile(file);
                 this.utf8 = zc.isUTF8();
             }
 
+            @SuppressWarnings("removal")
+            private static File getCanonicalFile(File file) throws IOException {
+                if (System.getSecurityManager() == null) {
+                    return file.getCanonicalFile();
+                } else {
+                    try {
+                        return AccessController.doPrivileged(
+                                (PrivilegedExceptionAction<File>) file::getCanonicalFile);
+                    } catch (PrivilegedActionException pae) {
+                        throw new IOException(pae.getException());
```

Comment:
The exception handling isn't right, you should be re-throwing pae.getException when it's an IOException, this should do what you want
```
        PrivilegedExceptionAction<File> pa = file::getCanonicalFile;
        try {
            return AccessController.doPrivileged(pa);
        } catch (PrivilegedActionException pae) {
            Throwable cause = pae.getCause();
            if (cause instanceof IOException ioe) {
                throw ioe;
            } else {
                throw new IOException(cause);
            }
        }
```